File: TTPlace/Tool/TTBox/UI/Components/TTCarousel/TTCarousel.m
//
//  TTCarousel.m
//  iCarouselExample
//
//  Created by Mr.hong on 2021/1/4.
//

#import "TTCarousel.h"
#import "UIImageView+WebCache.h"
#import "TTCarouselPageControl.h"


@interface TTCarousel()<iCarouselDataSource,iCarouselDelegate>

/**
 itemSize
 */
@property(nonatomic, readwrite, assign)CGSize itemSize;

@end

@implementation TTCarousel


// 初始化添加pageControll
- (instancetype)initWithItems:(NSArray *)items itemSize:(CGSize)itemSize
{
    self = [super initWithFrame:CGRectZero];
    if (self) {
        self.layer.masksToBounds = true;
        
        self.items = [items mutableCopy];
        self.itemSize = itemSize;
        
        self.dataSource = self;
        self.delegate = self;
        
        
//        TTCarouselPageControl *pageControl = [[TTCarouselPageControl alloc] initWithPageCount:items.count itemSize:CGSizeMake(5, 5)];
//        pageControl.backgroundColor = [UIColor redColor];
//        pageControl.currentPageIndicatorTintColor = [UIColor whiteColor];
//        pageControl.pageIndicatorTintColor = [UIColor grayColor];
//        [self addSubview:pageControl];
//        pageControl.type = iCarouselTypeLinear;
//        [pageControl mas_makeConstraints:^(MASConstraintMaker *make) {
//            make.top.mas_equalTo(carousel1.mas_bottom);
//            make.left.right.equalTo(carousel1);
//            make.height.offset(20);
//        }];
        
        
//        self.pageControl = [[TTCarousel alloc] init];
//        self.pageControl.hidden = true;
//        self.pageControl.currentPage = 0;
//        self.pageControl.numberOfPages = 3;
//        self.pageControl.pageIndicatorTintColor = [UIColor redColor];
//        self.pageControl.currentPageIndicatorTintColor = [UIColor yellowColor];
//        [self addSubview:self.pageControl];
        
        
        
        
        // layout
//        [self.pageControl mas_makeConstraints:^(MASConstraintMaker *make) {
//            make.left.right.offset(0);
//            make.center.mas_equalTo(self);
//            make.bottom.offset(0);
//            make.height.offset(20);
//        }];
        
        
        
        
        
        
        
        
        [self reloadData];
    }
    return self;
}

#pragma mark - delegate



- (NSInteger)numberOfItemsInCarousel:(iCarousel *)carousel {
    return (NSInteger)self.items.count;
}



- (UIView *)carousel:(iCarousel *)carousel viewForItemAtIndex:(NSInteger)index reusingView:(nullable UIView *)view {
    UIView *_view;
    if (view == nil) {
        _view = [[UIView alloc] init];
        _view.frame = CGRectMake(0, 0, self.itemSize.width, self.itemSize.height);
        UIImageView *imageView = [[UIImageView alloc] init];
        [_view addSubview:imageView];
        imageView.frame = view.frame;
          
        // 图片加载
        NSString *urlStr = self.items[index];
        [imageView sd_setImageWithURL:[NSURL URLWithString:urlStr]];
    }else {
        _view = view;
    }
    return _view;
}

- (CGFloat)carousel:(__unused iCarousel *)carousel valueForOption:(iCarouselOption)option withDefault:(CGFloat)value
{
    //customize carousel display
    switch (option)
    {
        case iCarouselOptionWrap:
        {
            //normally you would hard-code this to YES or NO
            return YES;
        }
        case iCarouselOptionSpacing:
        {
            //add a bit of spacing between the item views
            return value;
        }
        case iCarouselOptionFadeMax:
        {
//            if (self.carousel.type == iCarouselTypeCustom)
//            {
//                //set opacity based on distance from camera
//                return 0.0;
//            }
//            return value;
        }
        case iCarouselOptionShowBackfaces:
        case iCarouselOptionRadius:
        case iCarouselOptionAngle:
        case iCarouselOptionArc:
        case iCarouselOptionTilt:
        case iCarouselOptionCount:
        case iCarouselOptionFadeMin:
        case iCarouselOptionFadeMinAlpha:
        case iCarouselOptionFadeRange:
        case iCarouselOptionOffsetMultiplier:
        case iCarouselOptionVisibleItems:
        {
            return value;
        }
    }
}

#pragma mark -
#pragma mark iCarousel taps

- (void)carousel:(__unused iCarousel *)carousel didSelectItemAtIndex:(NSInteger)index
{
//    NSNumber *item = (self.items)[(NSUInteger)index];
    
//    NSLog(@"Tapped view number: %@", item);
    
    
    self.itemDidSelected(index);
}

- (void)carouselCurrentItemIndexDidChange:(__unused iCarousel *)carousel
{
    NSLog(@"Index: %@", @(self.currentItemIndex));
//    self.indexDidChange(self.currentItemIndex);
    
//    self.pageControl.currentPage = self.currentItemIndex;
}

- (void)carouselDidEndDecelerating:(iCarousel *)carousel {
    NSLog(@"333");
    self.indexDidChange(carousel.currentItemIndex);
}

- (void)carouselDidEndScrollingAnimation:(iCarousel *)carousel {
    NSLog(@"2222");
    self.indexDidChange(carousel.currentItemIndex);
}

- (void)carouselDidEndDragging:(iCarousel *)carousel willDecelerate:(BOOL)decelerate {
    self.indexDidChange(carousel.currentItemIndex);
    NSLog(@"1111");
}



@end

File: 新项目模板/App/TTBox/UI/Components/TTCarousel/TTCarousel.swift
//
//  TTCarousel.swift
//  Yuhun
//
//  Created by Mr.hong on 2021/2/5.
//

import Foundation


class TTCarousel: iCarousel, iCarouselDataSource, iCarouselDelegate {
    /// 当前下标改变block
    var indexDidChange: ((_ index: Int) -> Void)?
    
    /// 选中当前下标block
    var itemDidSelected: ((_ index: Int) -> Void)?
    
    /// 轮播图数据源
    var items: [String]?
    
    var itemSize = CGSize.zero

    
    init(items: [String]?, itemSize: CGSize) {
        super.init(frame: .zero)
        self.layer.masksToBounds = true
        
        self.items = items
        self.itemSize = itemSize
        
        self.dataSource = self
        self.delegate = self
        self.reloadData()
    }
    
    
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}


extension TTCarousel {
    func numberOfItems(in carousel: iCarousel) -> Int {
        return self.items?.count ?? 0
    }
    
    func carousel(_ carousel: iCarousel, viewForItemAt index: Int, reusing view: UIView?) -> UIView {
//        if let _view = view {
//           return _view
//        }else {
//           let _view = UIView()
//            _view?.frame = CGRect(x: 0, y: 0, width: itemSize.width, height: itemSize.height)
//            let imageView = UIImageView()
//            _view?.addSubview(imageView)
//            imageView.frame = view.frame
//
//            // 图片加载
//            let urlStr = items[index] as? String
//            imageView.sd_setImage(with: URL(string: urlStr ?? ""))
//            return _view
//        }
        return View()
    }
    
    func carousel(_ carousel: iCarousel, valueFor option: iCarouselOption, withDefault value: CGFloat) -> CGFloat {
        //customize carousel display
        switch (option)
        {
        case .wrap:
                //normally you would hard-code this to YES or NO
                return 0;
        case .spacing:
                //add a bit of spacing between the item views
                return value;
        default:
            return value;
            break
        }
    }
    
    // 点击事件
    func carousel(_ carousel: iCarousel, didSelectItemAt index: Int) {
        itemDidSelected?(index)
    }
    
    // 当先下标变更
    func carouselCurrentItemIndexDidChange(_ carousel: iCarousel) {
        print(self.currentItemIndex)
    }
    
    
    func carouselDidEndDecelerating(_ carousel: iCarousel) {
        indexDidChange?(carousel.currentItemIndex)
    }
    
    func carouselDidEndScrollingAnimation(_ carousel: iCarousel) {
        indexDidChange?(carousel.currentItemIndex)
    }
    
    func carouselDidEndDragging(_ carousel: iCarousel, willDecelerate decelerate: Bool) {
        indexDidChange?(carousel.currentItemIndex)
    }
}
